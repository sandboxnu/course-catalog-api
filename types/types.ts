/*
 * This file is part of Search NEU and licensed under AGPL3.
 * See the license file in the root folder for details.
 */

// A block of meetings, ex: "Tuesdays+Fridays, 9:50-11:30am"
import { ParsedTermSR } from "./scraperTypes";
import { Prisma } from "@prisma/client";

export interface BackendMeeting {
  startDate: number;
  endDate: number;
  where: string;
  type: string;
  times: Partial<
    Record<"0" | "1" | "2" | "3" | "4" | "5" | "6", MeetingTime[]>
  >;
}

// A single meeting time, ex: "9:50-11:30am"
export interface MeetingTime {
  start: number | string;
  end: number | string;
}

export interface FacultyMeetingTime {
  beginTime: string;
  building: string;
  buildingDescription: string;
  campus: string;
  campusDescription: string;
  category: string;
  class: string;
  courseReferenceNumber: string;
  creditHourSession: number;
  endDate: string;
  endTime: string;
  friday: boolean;
  hoursWeek: number;
  meetingScheduleType: string;
  meetingType: string;
  meetingTypeDescription: string;
  monday: boolean;
  room: string;
  saturday: boolean;
  startDate: string;
  sunday: boolean;
  term: string;
  thursday: boolean;
  tuesday: boolean;
  wednesday: boolean;
}

export const NEU_COLLEGE = "NEU";
export const CPS_COLLEGE = "CPS";
export const LAW_COLLEGE = "LAW";

export type CollegeNames = "NEU" | "CPS" | "LAW";

export interface EmployeeRequestResponse {
  RecordNumber: string;
  LastName: string;
  FirstName: string;
  LongPositionTitle?: string;
  PositionTitle?: string;
  Department: string;
  PhoneNumber: string;
  CampusAddress: string;
  Email: string;
  SecondaryDivision: string;
  SecondaryDepartment: string;
  PreferredFirstName: string;
}

export interface Employee {
  id: string;
  name: string;
  firstName: string;
  lastName: string;
  primaryDepartment?: string;
  primaryRole?: string;
  phone?: string;
  email?: string;
  officeRoom?: string;
}

// A course within a semester
export interface Course {
  host: string;
  termId: string;
  subject: string;
  classId: string;
  classAttributes: string[];
  desc: string;
  prettyUrl: string;
  name: string;
  url: string;
  lastUpdateTime: number;
  maxCredits: number;
  minCredits: number;
  coreqs: Requisite;
  prereqs: Requisite;
  feeAmount: number;
  feeDescription: string;
  sections?: Section[];
}

// A co or pre requisite object.
export type Requisite = string | BooleanReq | CourseReq;

export interface BooleanReq {
  type: "and" | "or";
  values: Requisite[];
}

export interface CourseReq {
  classId: string;
  subject: string;
  missing?: true;
}

export function isBooleanReq(req: Requisite): req is BooleanReq {
  return (req as BooleanReq).type !== undefined;
}

export function isCourseReq(req: Requisite): req is CourseReq {
  return (req as CourseReq).classId !== undefined;
}

// A section of a course
export interface Section {
  host: string;
  termId: string;
  subject: string;
  classId: string;
  classType: string;
  crn: string;
  seatsCapacity: number;
  seatsRemaining: number;
  waitCapacity: number;
  waitRemaining: number;
  lastUpdateTime: number;
  campus: string;
  honors: boolean;
  url: string;
  profs: string[];
  meetings: BackendMeeting[];
}

export interface TermInfo {
  termId: string;
  subCollege: string;
  text: string;
}

export interface CourseRef {
  subject: string;
  termId: string;
  classId: string;
}

export interface Dump {
  termDump?: ParsedTermSR;
  profDump?: Employee[];
  destroy?: boolean;
  currentTermInfos?: TermInfo[];
}

export type BulkUpsertInput =
  | Prisma.SectionCreateInput
  | Prisma.CourseCreateInput
  | Prisma.ProfessorCreateInput;

export type SingleTransformFunction = (any) => string;
export type ArrayTransformFunction = (
  any,
  string,
  SingleTransformFunction
) => string;
export type TransformFunction =
  | SingleTransformFunction
  | ArrayTransformFunction;

export type EmptyObject = Record<string, never>;
