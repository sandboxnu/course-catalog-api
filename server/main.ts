import graphQlServer from "./graphql/server";
import macros from "../utils/macros";
import notificationServer from "./notifs/server";

const port = 8080;
notificationServer.listen(port, () => {
  console.log("Running twilio notification server on port %s", port);
});

graphQlServer
  .listen()
  .then(({ url }) => {
    macros.log(`ready at ${url}`);
    return;
  })
  .catch((err) => {
    macros.error(`error starting graphql server: ${JSON.stringify(err)}`);
  });
