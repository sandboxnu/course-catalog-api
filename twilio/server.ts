import express from "express";
import cors from "cors";
import { createServer } from "http";
import jwt from "jsonwebtoken";
import {
  sendVerificationCode,
  checkVerificationCode,
  handleUserReply,
} from "./notifs";

const corsOptions = {
  origin: process.env.CLIENT_ORIGIN,
};

const app = express();
app.use(cors(corsOptions));
const port = 8080;
app.use(express.json());
const server = createServer(app);

server.listen(port, () => {
  console.log("Running twilio notification server on port %s", port);
});

app.get("/knockknock", (req, res) => res.send("Who's there?"));

app.post("/twilio/sms", (req, res) => handleUserReply(req, res));

app.post("/sms/signup", (req, res) => {
  // twilio needs the phone number in E.164 format see https://www.twilio.com/docs/verify/api/verification
  const phoneNumber = req.body.phoneNumber;
  if (!phoneNumber) {
    res.status(400).send("Missing phone number.");
  }
  sendVerificationCode(phoneNumber)
    .then((response) => {
      res.status(response.statusCode).send(response.message);
    })
    .catch((e) =>
      res.status(500).send("Error trying to send verification code")
    );
});

app.post("/sms/verify", (req, res) => {
  const phoneNumber = req.body.phoneNumber;
  const verificationCode = req.body.verificationCode;
  if (!phoneNumber || !verificationCode) {
    return res.status(400).send("Missing phone number or verification code.");
  }

  checkVerificationCode(phoneNumber, verificationCode)
    .then((response) => {
      if (response.statusCode === 200) {
        const token = jwt.sign({ phoneNumber }, process.env.JWT_SECRET);
        res
          .status(response.statusCode)
          .send({ message: response.message, token });
      } else {
        res.status(response.statusCode).send(response.message);
      }
    })
    .catch((e) => res.status(500).send("Error trying to verify code"));
});
