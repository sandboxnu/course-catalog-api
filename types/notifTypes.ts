export interface UserInfo {
  phoneNumber: string;
  courseIds: string[];
  sectionIds: string[];
}

// Stores information for all changes to a course or section
export interface NotificationInfo {
  updatedCourses: CourseNotificationInfo[];
  updatedSections: SectionNotificationInfo[];
}

// marks new sections being added to a Course
export interface CourseNotificationInfo {
  subject: string;
  courseId: string;
  termId: string;
  courseHash: string;
  numberOfSectionsAdded: number;
  campus: string;
}

// marks seats becoming available in a section
export interface SectionNotificationInfo {
  subject: string;
  courseId: string;
  sectionHash: string;
  termId: string;
  seatsRemaining: number;
  crn: string;
  campus: string;
}
