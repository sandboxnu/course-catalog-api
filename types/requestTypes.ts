import { CoreOptions } from "request";
import http from "http";
import https from "https";

export interface HostAnalytics {
  totalBytesDownloaded: number;
  totalErrors: number;
  totalGoodRequests: number;
  startTime: number | null;
}

export type RequestAnalytics = Record<string, HostAnalytics>;

export interface RequestPool {
  maxSockets: number;
  keepAlive: boolean;
  maxFreeSockets: number;
  "https:false:ALL"?: https.Agent;
  "http:"?: http.Agent;
}

export interface AgentAnalytics {
  socketCount: number;
  requestCount: number;
  maxSockets: number;
}

export interface CustomRequestConfig extends CoreOptions {
  url: string;
  method: string;
  cache?: boolean;
  cacheName?: string;
  headers?: Record<string, string>;
  pool?: RequestPool;
}

export type PartialRequestConfig =
  | (Partial<CustomRequestConfig> & { url: string })
  | string;

export interface NativeRequestConfig extends CustomRequestConfig {
  method: string;
  headers: Record<string, string>;
  pool: RequestPool;
  timeout: number;
  resolveWithFullResponse: boolean;
  rejectUnauthorized: boolean;
  requestCert: boolean;
  ciphers: string;
}

export type AmplitudeEvent = Partial<AgentAnalytics> & {
  totalBytesDownloaded?: number;
  totalErrors?: number;
  totalGoodRequests?: number;
  startTime?: number | null;
  hostname: string;
};

export type DoRequestReturn = { items: unknown; totalCount: number };
