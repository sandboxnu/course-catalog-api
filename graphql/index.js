import { ApolloServer, gql } from "apollo-server";
import GraphQLJSON, { GraphQLJSONObject } from "graphql-type-json";
import macros from "../utils/macros";

import employeeTypeDef from "./typeDefs/employee";

import searchResolvers from "./resolvers/search";
import searchTypeDef from "./typeDefs/search";

import classResolvers from "./resolvers/class";
import classTypeDef from "./typeDefs/class";
import classOccurrenceTypeDef from "./typeDefs/classOccurrence";

import majorResolvers from "./resolvers/major";
import majorTypeDef from "./typeDefs/major";
import majorOccurrenceTypeDef from "./typeDefs/majorOccurrence";

if (macros.PROD || process.env.ENABLE_NOTIFS) {
  require("../twilio/server");
}

// Enable JSON custom type
const JSONResolvers = {
  JSON: GraphQLJSON,
  JSONObject: GraphQLJSONObject,
};

// Base query so other typeDefs can do "extend type Query"
const baseQuery = gql`
  scalar JSON
  scalar JSONObject

  type Query {
    _empty: String
  }

  type PageInfo {
    # When paginating forwards, are there more items
    hasNextPage: Boolean!
  }
`;

const server = new ApolloServer({
  typeDefs: [
    baseQuery,
    classTypeDef,
    classOccurrenceTypeDef,
    employeeTypeDef,
    majorTypeDef,
    majorOccurrenceTypeDef,
    searchTypeDef,
  ],
  resolvers: [JSONResolvers, classResolvers, majorResolvers, searchResolvers],
});

if (require.main === module) {
  server
    .listen()
    .then(({ url }) => {
      macros.log(`ready at ${url}`);
      return;
    })
    .catch((err) => {
      macros.error(`error starting graphql server: ${JSON.stringify(err)}`);
    });
}

export default server;
