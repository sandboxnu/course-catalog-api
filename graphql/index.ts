import { ApolloServer, gql } from "apollo-server";
import GraphQLJSON, { GraphQLJSONObject } from "graphql-type-json";
import macros from "../utils/macros";
import prisma from "../services/prisma";

import employeeTypeDef from "./typeDefs/employee";

import searchResolvers from "./resolvers/search";
import searchTypeDef from "./typeDefs/search";

import classResolvers from "./resolvers/class";
import classTypeDef from "./typeDefs/class";
import classOccurrenceTypeDef from "./typeDefs/classOccurrence";

import majorResolvers from "./resolvers/major";
import majorTypeDef from "./typeDefs/major";
import majorOccurrenceTypeDef from "./typeDefs/majorOccurrence";

import termInfoResolvers from "./resolvers/termInfo";
import termInfoTypeDef from "./typeDefs/termInfo";

if (macros.PROD || process.env.ENABLE_NOTIFS) {
  require("../twilio/server");
}

// Enable JSON custom type
const JSONResolvers = {
  JSON: GraphQLJSON,
  JSONObject: GraphQLJSONObject,
};

// Base query so other typeDefs can do "extend type Query"
const baseQuery = gql`
  scalar JSON
  scalar JSONObject

  type Query {
    _empty: String
  }

  type PageInfo {
    # When paginating forwards, are there more items
    hasNextPage: Boolean!
  }
`;

const server = new ApolloServer({
  typeDefs: [
    baseQuery,
    classTypeDef,
    classOccurrenceTypeDef,
    employeeTypeDef,
    majorTypeDef,
    majorOccurrenceTypeDef,
    searchTypeDef,
    termInfoTypeDef,
  ],
  resolvers: [
    JSONResolvers,
    classResolvers,
    majorResolvers,
    searchResolvers,
    termInfoResolvers,
  ],
});

if (require.main === module) {
  server
    .listen()
    .then(({ url }) => {
      macros.log(`ready at ${url}`);
      return;
    })
    .catch((err) => {
      macros.error(`error starting graphql server: ${JSON.stringify(err)}`);
    });

  // TEMP / TODO - REMOVE / DO NOT LEAVE HERE PLEASE
  // Temp fix to address Prisma connection pool issues
  // https://github.com/prisma/prisma/issues/7249#issuecomment-1059719644
  const intervalTime = 6 * 60 * 60_000; // Every 6 hours

  if (!macros.TEST) {
    setInterval(async () => {
      const startTime = Date.now();
      await prisma.$disconnect();
      macros.log("Disconnected Prisma");
      await prisma.$connect();
      const totalTime = Date.now() - startTime;
      macros.log(
        `Reconnected Prisma - downtime of ${totalTime} ms (${
          totalTime / 60_000
        } mins)`
      );
    }, intervalTime);
  }
}

export default server;
