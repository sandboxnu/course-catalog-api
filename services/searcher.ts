/*
 * This file is part of Search NEU and licensed under AGPL3.
 * See the license file in the root folder for details.
 */
import _ from "lodash";
import {
  Course as PrismaCourse,
  Section as PrismaSection,
} from "@prisma/client";
import prisma from "../services/prisma";
import elastic, { Elastic } from "../utils/elastic";
import HydrateSerializer from "../serializers/hydrateSerializer";
import macros from "../utils/macros";
import {
  EsQuery,
  QueryNode,
  ExistsQuery,
  TermsQuery,
  TermQuery,
  LeafQuery,
  MATCH_ALL_QUERY,
  RangeQuery,
  EsFilterStruct,
  EsAggFilterStruct,
  FilterInput,
  FilterPrelude,
  AggFilterPrelude,
  SortInfo,
  Range,
  SearchResults,
  SingleSearchResult,
  PartialResults,
  AggCount,
  EsResultBody,
  EsMultiResult,
  AggResults,
  SearchResult,
  CourseSearchResult,
  ParsedQuery,
} from "../types/searchTypes";
import { SerializedCourse } from "../types/serializerTypes";
import { Course, Section } from "../types/types";

type CourseWithSections = PrismaCourse & { sections: PrismaSection[] };
type SSRSerializerOutput = { [id: string]: CourseSearchResult };

class Searcher {
  elastic: Elastic;

  subjects: Record<string, string>;

  filters: FilterPrelude;

  aggFilters: AggFilterPrelude;

  AGG_RES_SIZE: number;

  /**
   * Regex for matching a course-code (eg. "CS3500")
   */
  COURSE_CODE_PATTERN: RegExp;

  constructor() {
    this.elastic = elastic;
    this.subjects = {};
    this.filters = Searcher.generateFilters();
    this.aggFilters = _.pickBy<EsFilterStruct, EsAggFilterStruct>(
      this.filters,
      (f): f is EsAggFilterStruct => f.agg !== false,
    );
    this.AGG_RES_SIZE = 1000;
    this.COURSE_CODE_PATTERN = /^\s*([a-zA-Z]{2,4})\s*(\d{4})?\s*$/i;
  }

  static generateFilters(): FilterPrelude {
    // type validating functions
    const isString = (arg: unknown): arg is string => {
      return typeof arg === "string";
    };

    const isBoolean = (arg: unknown): arg is boolean => {
      return typeof arg === "boolean";
    };

    const isStringArray = (arg: unknown): arg is string[] => {
      return Array.isArray(arg) && arg.every((elem) => isString(elem));
    };

    const isTrue = (arg: unknown): arg is true => {
      return typeof arg === "boolean" && arg;
    };

    const isNum = (arg: unknown): arg is number => {
      return typeof arg === "number";
    };

    const isRange = (arg: any): arg is Range => {
      return (
        _.difference(Object.keys(arg), ["min", "max"]).length === 0 &&
        isNum(arg.min) &&
        isNum(arg.max)
      );
    };

    // filter-generating functions
    const getSectionsAvailableFilter = (): ExistsQuery => {
      return { exists: { field: "sections" } };
    };

    const getNUpathFilter = (selectedNUpaths: string[]): TermsQuery => {
      return { terms: { "class.nupath.keyword": selectedNUpaths } };
    };

    const getSubjectFilter = (selectedSubjects: string[]): TermsQuery => {
      return { terms: { "class.subject.keyword": selectedSubjects } };
    };

    const getClassTypeFilter = (selectedClassTypes: string[]): TermsQuery => {
      return { terms: { "sections.classType.keyword": selectedClassTypes } };
    };

    const getTermIdFilter = (selectedTermId: string): TermQuery => {
      return { term: { "class.termId": selectedTermId } };
    };

    const getRangeFilter = (selectedRange: Range): RangeQuery => {
      return {
        range: {
          "class.classId.numeric": {
            gte: selectedRange.min,
            lte: selectedRange.max,
          },
        },
      };
    };

    const getCampusFilter = (selectedCampuses: string[]): TermsQuery => {
      return { terms: { "sections.campus.keyword": selectedCampuses } };
    };

    const getHonorsFilter = (selectedHonors: boolean): TermQuery => {
      return { term: { "sections.honors": selectedHonors } };
    };

    return {
      nupath: {
        validate: isStringArray,
        create: getNUpathFilter,
        agg: "class.nupath.keyword",
      } as EsFilterStruct<string[]>,
      subject: {
        validate: isStringArray,
        create: getSubjectFilter,
        agg: "class.subject.keyword",
      } as EsFilterStruct<string[]>,
      classType: {
        validate: isStringArray,
        create: getClassTypeFilter,
        agg: "sections.classType.keyword",
      } as EsFilterStruct<string[]>,
      sectionsAvailable: {
        validate: isTrue,
        create: getSectionsAvailableFilter,
        agg: false,
      } as EsFilterStruct<true>,
      classIdRange: {
        validate: isRange,
        create: getRangeFilter,
        agg: false,
      } as EsFilterStruct<Range>,
      termId: {
        validate: isString,
        create: getTermIdFilter,
        agg: false,
      } as EsFilterStruct<string>,
      campus: {
        validate: isStringArray,
        create: getCampusFilter,
        agg: "sections.campus.keyword",
      } as EsFilterStruct<string[]>,
      honors: {
        validate: isBoolean,
        create: getHonorsFilter,
        agg: "sections.honors",
      } as EsFilterStruct<boolean>,
    };
  }

  async initializeSubjects(): Promise<void> {
    if (_.isEmpty(this.subjects)) {
      (await prisma.subject.findMany()).forEach((obj) => {
        this.subjects[obj.abbreviation] = obj.description;
      });
    }
  }

  /**
   * Remove any invalid filter with the following criteria:
   * 1. Correct key string and value type;
   * 2. Check that { online: false } should never be in filters
   *
   * A sample filters JSON object has the following format:
   * { 'nupath': string[],
   *   'college': string[],
   *   'subject': string[],
   *   'classType': string }
   *
   * @param {object} filters The json object represting all filters on classes
   */
  validateFilters(filters: FilterInput): FilterInput {
    const validFilters: FilterInput = {};
    Object.keys(filters).forEach((currFilter) => {
      if (!(currFilter in this.filters)) {
        macros.warn("Invalid filter key.", currFilter);
      } else if (!this.filters[currFilter].validate(filters[currFilter])) {
        macros.warn("Invalid filter value type.", currFilter);
      } else {
        validFilters[currFilter] = filters[currFilter];
      }
    });
    return validFilters;
  }

  getFields(): string[] {
    return [
      "class.name^2", // Boost by 2
      "class.name.autocomplete",
      "class.subject^4",
      "class.classId^3",
      "sections.profs",
      "sections.crn",
      "employee.name^2",
      "employee.email",
      "employee.phone",
    ];
  }

  /**
   * Given a string, creates a list of queries that are either phrase queries or
   * most_field queries. Phrases are between quotes, and use phrase matching.
   * Most_field queries are everything else, and search the terms on all fields.
   * @param query the string that is parsed into a list
   * @returns an object containing a list of phrase_queries, and a field query.
   */
  parseQuery(query: string): ParsedQuery {
    const matches = [...query.matchAll(/"(.*?)"/gi)];
    const matchedPhrases = matches.map((match) => {
      return match[0];
    });

    const matchesRegExp = new RegExp(matchedPhrases.join("|"), "gi");

    //make sure theres no extra white space after removing the phrases.
    const nonMatches = query
      .replace(matchesRegExp, "")
      .replace('"', "")
      .replace(/\s+/g, " ")
      .trim();

    // go through the phrases, and make phrase queries with them.
    const phraseQueries = matches.map((match) => {
      return {
        multi_match: {
          query: match[1],
          type: "phrase",
          fields: this.getFields(),
        },
      };
    });

    //make the field query and add it to the list.
    let fieldQuery: LeafQuery = null;
    if (nonMatches.trim() !== "") {
      fieldQuery = {
        multi_match: {
          query: nonMatches,
          type: "most_fields",
          fields: this.getFields(),
        },
      };
    }

    return {
      phraseQ: phraseQueries,
      fieldQ: fieldQuery,
    };
  }

  /**
   * Get elasticsearch query
   */
  generateQuery(
    query: string,
    termId: string,
    userFilters: FilterInput,
    min: number,
    max: number,
    aggregation = "",
  ): EsQuery {
    //a list of all queries
    const matchQueries: ParsedQuery = this.parseQuery(query);

    const phraseQueries: LeafQuery[] = matchQueries.phraseQ;

    const fieldQuery: LeafQuery = matchQueries.fieldQ;
    const fieldQueryExists = fieldQuery !== null;

    // use lower classId has tiebreaker after relevance
    const sortByClassId: SortInfo = {
      "class.classId.keyword": { order: "asc", unmapped_type: "keyword" },
    };

    // filter by type employee
    const isEmployee: LeafQuery = { term: { type: "employee" } };
    const areFiltersApplied: boolean = Object.keys(userFilters).length > 0;
    const requiredFilters: FilterInput = {
      termId: termId,
      sectionsAvailable: true,
    };
    const filters: FilterInput = { ...requiredFilters, ...userFilters };

    const classFilters: QueryNode[] = _(filters)
      .pick(Object.keys(this.filters))
      .toPairs()
      .map(([key, val]) => this.filters[key].create(val))
      .value();

    const aggQuery = !aggregation
      ? undefined
      : {
          [aggregation]: {
            terms: {
              field: this.aggFilters[aggregation].agg,
              size: this.AGG_RES_SIZE,
            },
          },
        };

    // compound query for text query and filters
    return {
      from: min,
      size: max - min,
      sort: ["_score", sortByClassId],
      query: {
        bool: {
          must:
            phraseQueries.length === 0 && !fieldQueryExists
              ? MATCH_ALL_QUERY
              : phraseQueries,
          filter: {
            bool: {
              should: [
                { bool: { must: classFilters } },
                ...(!areFiltersApplied ? [isEmployee] : []),
              ],
            },
          },
          should: fieldQuery, //should match any remaining terms
          minimum_should_match: 0,
        },
      },
      aggregations: aggQuery,
    };
  }

  generateMQuery(
    query: string,
    termId: string,
    min: number,
    max: number,
    filters: FilterInput,
  ): EsQuery[] {
    const validFilters: FilterInput = this.validateFilters(filters);

    const queries: EsQuery[] = [
      this.generateQuery(query, termId, validFilters, min, max),
    ];

    for (const fKey of Object.keys(this.aggFilters)) {
      const everyOtherFilter: FilterInput = _.omit(filters, fKey);
      queries.push(
        this.generateQuery(query, termId, everyOtherFilter, 0, 0, fKey),
      );
    }
    return queries;
  }

  async getSearchResults(
    query: string,
    termId: string,
    min: number,
    max: number,
    filters: FilterInput,
  ): Promise<PartialResults> {
    const queries = this.generateMQuery(query, termId, min, max, filters);
    const results: EsMultiResult = await elastic.mquery(
      `${elastic.CLASS_ALIAS},${elastic.EMPLOYEE_ALIAS}`,
      queries,
    );
    return this.parseResults(
      results.body.responses,
      Object.keys(this.aggFilters),
    );
  }

  parseResults(results: EsResultBody[], filters: string[]): PartialResults {
    return {
      output: results[0].hits.hits,
      resultCount: results[0].hits.total.value,
      took: results[0].took,
      aggregations: _.fromPairs(
        filters.map((filter, idx) => {
          return [
            filter,
            results[idx + 1].aggregations[filter].buckets.map((aggVal) => {
              return this.generateAgg(filter, aggVal.key, aggVal.doc_count);
            }),
          ];
        }),
      ),
    };
  }

  filterOutSections(
    results: SearchResult[],
    filters: FilterInput,
  ): SearchResult[] {
    return results
      .map((result) => {
        if (result.type === "employee") {
          return result;
        }

        result.sections = result.sections.filter((section) => {
          return Object.keys(filters).every((filter) => {
            const filterValue = filters[filter];
            switch (filter) {
              case "honors":
                return section.honors === filterValue;
              case "subject":
                return (filterValue as string[]).includes(section.subject);
              case "campus":
                return (filterValue as string[]).includes(section.campus);
              case "classType":
                return (filterValue as string[]).includes(section.classType);
              default:
                return true;
            }
          });
        });

        return result.sections.length > 0 ? result : null;
      })
      .filter((result) => result !== null) as SearchResult[];
  }

  generateAgg(filter: string, value: string, count: number): AggCount {
    const agg: AggCount = { value, count };
    // in addition to the subject abbreviation, add subject description for subject filter
    if (filter === "subject") {
      agg.description = this.subjects[value];
    }
    return agg;
  }

  async getOneSearchResult(
    subject: string,
    classId: string,
    termId: string,
  ): Promise<SingleSearchResult> {
    const start = Date.now();

    const results = await this.getSearchResults(
      subject + classId,
      termId,
      0,
      9999,
      {
        subject: [subject],
      },
    );
    const result = results?.output[0];

    const hasSections = result?._source?.sections?.length > 0;
    const subjectMatches = result?._source?.class?.subject === subject;
    const codeMatches = result?._source?.class?.classId === classId;
    // Only show this course
    const showCourse = hasSections && subjectMatches && codeMatches;

    let aggregations: AggResults;
    let resultOutput: SearchResult[];

    if (showCourse) {
      resultOutput = await new HydrateSerializer().bulkSerialize([result]);

      aggregations = this.getSingleResultAggs({
        ...result._source?.class,
        sections: result._source?.sections,
      });
    }

    return {
      results: resultOutput ?? [],
      resultCount: showCourse ? 1 : 0,
      took: 0,
      hydrateDuration: Date.now() - start,
      aggregations: aggregations ?? {
        nupath: [],
        subject: [],
        classType: [],
        campus: [],
        honors: [],
      },
    };
  }

  getSingleResultAggs(result: CourseWithSections): AggResults {
    return {
      nupath: result.nupath.map((val) => {
        return { value: val, count: 1 };
      }),
      subject: [this.generateAgg("subject", result.subject, 1)],
      classType: [{ value: result.sections[0].classType, count: 1 }],
      campus: [{ value: result.sections[0].campus, count: 1 }],
      honors: [{ value: result.sections[0].honors?.toString(), count: 1 }],
    };
  }

  /**
   * Search for classes and employees
   * @param  {string}  query  The search to query for
   * @param  {string}  termId The termId to look within
   * @param  {number} min    The index of first document to retreive
   * @param  {number} max    The index of last document to retreive
   * @param filters
   */
  async search(
    query: string,
    termId: string,
    min: number,
    max: number,
    filters: FilterInput = {},
  ): Promise<SearchResults> {
    await this.initializeSubjects();
    const start = Date.now();

    let results: SearchResult[];
    let resultCount: number;
    let took: number;
    let hydrateDuration: number;
    let aggregations: AggResults;

    // if we know that the query is of the format of a course code, we want to return only one result
    const isSingleCourse = query.match(this.COURSE_CODE_PATTERN);

    const subject = isSingleCourse ? isSingleCourse[1].toUpperCase() : "";
    const isSubjectValid = subject in this.subjects;

    const courseCode = isSingleCourse ? isSingleCourse[2] : "";
    const isCourseCodeValid = macros.isNumeric(courseCode);

    if (isSingleCourse && isSubjectValid && isCourseCodeValid) {
      const singleResult = await this.getOneSearchResult(
        subject,
        courseCode,
        termId,
      );
      ({ results, resultCount, took, hydrateDuration, aggregations } =
        singleResult);
    } else {
      const searchResults = await this.getSearchResults(
        query,
        termId,
        min,
        max,
        filters,
      );
      ({ resultCount, took, aggregations } = searchResults);
      const startHydrate = Date.now();
      results = await new HydrateSerializer().bulkSerialize(
        searchResults.output,
      );
      results = this.filterOutSections(results, filters);
      hydrateDuration = Date.now() - startHydrate;
    }

    return {
      searchContent: results,
      resultCount,
      took: {
        total: Date.now() - start,
        hydrate: hydrateDuration,
        es: took,
      },
      aggregations,
    };
  }
}

const instance = new Searcher();
export default instance;
