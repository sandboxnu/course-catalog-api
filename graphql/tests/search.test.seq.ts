import { gql } from "apollo-server";
import { mocked } from "jest-mock";
import searcher from "../../services/searcher";
import server from "../index";
import { Course, Requisite, Section } from "../../types/types";
import { DocumentNode } from "graphql";
import { GraphQLResponse } from "apollo-server-core";

jest.mock("../../services/searcher");

const query = async (queryBody: {
  query: string | DocumentNode;
}): Promise<GraphQLResponse> => {
  return server.executeOperation(queryBody);
};

const EMPTY_REQ: Requisite = {
  type: "or",
  values: [],
};

const FUNDIES: Course = {
  classId: "2500",
  name: "Fundamentals of Computer Science 1",
  termId: "202030",
  subject: "CS",
  host: "neu.edu",
  classAttributes: [],
  prettyUrl: "pretty",
  desc: "a class",
  url: "url",
  lastUpdateTime: 20,
  maxCredits: 4,
  minCredits: 0,
  coreqs: EMPTY_REQ,
  prereqs: EMPTY_REQ,
  feeAmount: 0,
  feeDescription: "None",
};

const SECTION: Section = {
  host: "neu.edu",
  crn: "1234",
  classId: "2500",
  classType: "lecture",
  termId: "202030",
  subject: "CS",
  seatsCapacity: 1,
  seatsRemaining: 1,
  lastUpdateTime: 20,
  waitCapacity: 0,
  waitRemaining: 0,

  campus: "Boston",
  honors: false,
  url: "url",
  profs: [],
  meetings: [],
};

describe("search resolver", () => {
  beforeEach(() => {
    mocked(searcher).search.mockClear();
  });

  it("searches for blank in term gets results", async () => {
    mocked(searcher).search.mockResolvedValue({
      aggregations: {
        nupath: [{ value: "Writing Intensive", count: 10 }],
        classType: [{ value: "Lecture", count: 10 }],
      },
      resultCount: 10,
      searchContent: [
        {
          type: "class",
          class: FUNDIES,
          sections: [SECTION],
        },
      ],
      took: { total: 1, es: 1, hydrate: 0 },
    });
    const res = await query({
      query: gql`
        query search {
          search(termId: "202030") {
            totalCount
            pageInfo {
              hasNextPage
            }
            nodes {
              __typename
              ... on Employee {
                name
              }
              ... on ClassOccurrence {
                name
                classAttributes
                url
                lastUpdateTime
                desc
                prereqsFor
                optPrereqsFor
                sections {
                  profs
                  crn
                }
              }
            }
            filterOptions {
              nupath {
                value
                count
              }
              subject {
                value
                count
              }
              classType {
                value
                count
              }
              campus {
                value
                count
              }
            }
          }
        }
      `,
    });
    expect(mocked(searcher).search.mock.calls[0]).toEqual([
      "",
      "202030",
      0,
      10,
      {},
    ]);
    expect(res).toMatchSnapshot({});
  });

  it("searches with filter works", async () => {
    mocked(searcher).search.mockResolvedValue({
      aggregations: {
        nupath: [{ value: "Writing Intensive", count: 10 }],
        classType: [{ value: "Lecture", count: 10 }],
      },
      resultCount: 0,
      searchContent: [],
      took: { total: 1, es: 1, hydrate: 0 },
    });
    await query({
      query: gql`
        query search {
          search(
            termId: "202030"
            query: "hey"
            subject: ["CS"]
            nupath: ["bruh"]
            classIdRange: { min: 1000, max: 2500 }
          ) {
            totalCount
          }
        }
      `,
    });
    expect(mocked(searcher).search.mock.calls[0]).toEqual([
      "hey",
      "202030",
      0,
      10,
      {
        subject: ["CS"],
        nupath: ["bruh"],
        classIdRange: { min: 1000, max: 2500 },
      },
    ]);
  });

  it("shows hasnextpage false when no more", async () => {
    mocked(searcher).search.mockResolvedValue({
      aggregations: {},
      resultCount: 1,
      searchContent: [
        {
          type: "class",
          class: FUNDIES,
          sections: [SECTION],
        },
      ],
      took: { total: 1, es: 1, hydrate: 0 },
    });
    const res = await query({
      query: gql`
        query search {
          search(termId: "202030") {
            totalCount
            pageInfo {
              hasNextPage
            }
          }
        }
      `,
    });
    expect(mocked(searcher).search.mock.calls[0]).toEqual([
      "",
      "202030",
      0,
      10,
      {},
    ]);
    expect(res.data.search).toEqual({
      totalCount: 1,
      pageInfo: { hasNextPage: false },
    });
  });
});
