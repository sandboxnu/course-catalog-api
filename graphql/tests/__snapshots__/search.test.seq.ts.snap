// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search resolver searches for blank in term gets results 1`] = `
Object {
  "data": Object {
    "search": Object {
      "filterOptions": Object {
        "campus": null,
        "classType": Array [
          Object {
            "count": 10,
            "value": "Lecture",
          },
        ],
        "nupath": Array [
          Object {
            "count": 10,
            "value": "Writing Intensive",
          },
        ],
        "subject": null,
      },
      "nodes": Array [
        Object {
          "__typename": "ClassOccurrence",
          "classAttributes": Array [],
          "desc": "a class",
          "lastUpdateTime": 20,
          "name": "Fundamentals of Computer Science 1",
          "optPrereqsFor": null,
          "prereqsFor": null,
          "sections": Array [
            Object {
              "crn": "1234",
              "profs": Array [],
            },
          ],
          "url": "url",
        },
      ],
      "pageInfo": Object {
        "hasNextPage": false,
      },
      "totalCount": 10,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
